Module lifecycle note (#1964)
google/flax: docs/design_notes/module_lifecycle.rst

#. **The ``setup`` method**, which creates submodules and assigns them to attributes.
#. **User methods**. By convention, most Modules have just one  ``__call__`` method, but you can define multiple methods or use different method names.

Construction/initialization
=============================

Now we want to construct and use the ``MLP`` Module:

.. testcode::

  mlp = MLP(hidden_size=5, out_size=3)
  x = jax.numpy.ones((1, 2))
  variables = mlp.init(random.PRNGKey(0), x)
  y = mlp.apply(variables, x)

First, we construct an instance of ``MLP`` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The ``MLP`` constructor does not actually create variables or any internal state whatsoever. It's best to think of it as a specification or template of the Module that contains functionality but no data.

Let's take a closer look at initialization. Surprisingly, there actually is no seperate initialization path in Flax. Calling ``init`` is just a special case of ``apply``, which you can also write as:

> .. testcode:

Reviewer: I'm not seeing this rendered at:
 https://flax--1964.org.readthedocs.build/en/1964/design_notes/module_lifecycle.html
is there a subtle formatting error here that's accidentally hiding it?

Author: yup missing double colon. Ugh I'm not a fan of RestructuredText

The Module lifecycle
######################

.. testsetup::

  from typing import Any, Callable, Iterable
  import flax
  from flax import linen as nn
  from jax import random
  import jax

This design note is intended for users who are already familiar with linen Modules but want to understand more about the design principles behind the abstraction. This note should give you a good understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, check out the `MNIST Tutorial <https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html>`_.

> linen Modules offer a pythonic abstracton on top of Flax core. The ``Module`` abstraction allows you to create classes that have state, parameters and randomness on top of JAX. This is a practical guide to the design and behavior of the ``Module`` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.

Reviewer 2: Nit: "Pythonic" (capitalized)

```suggestion
linen Modules offer a Pythonic abstracton on top of Flax core. The ``Module`` abstraction allows you to create classes that have state, parameters and randomness on top of JAX. This is a practical guide to the design and behavior of the ``Module`` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.
```

Overview
***********

Definition
=============

> Let's start with a high-level overview of the Module lifecycle. We start by defining a simple Module:

Reviewer 2: Nit: to avoid repeating "start", maybe try

```suggestion
Let's start with a high-level overview of the Module lifecycle. First, define a simple Module:
```

      a = self.hidden(x)
      h = nn.relu(a)
      return self.out(h)

This Module consists of:

#. **Attribute annotations**, defined as `dataclass <https://docs.python.org/3/library/dataclasses.html>`_ fields. These annotations automatically define a constructor.
#. **The ``setup`` method**, which creates submodules and assigns them to attributes.
#. **User methods**. By convention, most Modules have just one  ``__call__`` method, but you can define multiple methods or use different method names.

Construction/initialization
=============================

Now we want to construct and use the ``MLP`` Module:

.. testcode::

  mlp = MLP(hidden_size=5, out_size=3)
  x = jax.numpy.ones((1, 2))
  variables = mlp.init(random.PRNGKey(0), x)
  y = mlp.apply(variables, x)

First, we construct an instance of ``MLP`` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The ``MLP`` constructor does not actually create variables or any internal state whatsoever. It's best to think of it as a specification or template of the Module that contains functionality but no data.

> Let's first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply`` and we could also code it as:

Reviewer 2: Nit: to avoid repeating "first" in 2 consecutive paragraphs, maybe try this 

```suggestion
Let's take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply``, which you can also write as:
```

And maybe try rephrasing the first or second "special" to avoid repeating "special" twice.

.. testcode:

  # equivalent to: variables = mlp.init(random.PRNGKey(0), x)
  _, variables = mlp.apply({}, x, rngs={"params": random.PRNGKey(0)}, mutable=True)

Thus, ``init`` is nothing more than a special case of ``apply`` where:

#. We call a Module without any initial variables (an empty dict).

> #. A PRNG generator named “params” is always passed for randomly initializing parameters (using the parameter initialization function).

Reviewer 2: Nit: maybe use backticks for code - e.g. ``params`` - instead of "params"

#. All variable collections are set to mutable (``mutable=True``). When a collection is mutable, existing variables can be updated and new variables can be created. Thus, inside ``init`` variables can be initialized in any variable collection and they are all added to the returned variable dictionary.

Lifecycle
=============

> Now that we understand ``init`` is just a special case of ``apply``, let's analyze what ``.apply(...)`` does in more detail. In fact, most of the complexity of Modules resides in the ``apply`` method. The "Module lifecycle" consists of constructing and ``apply``-ing a Module. We can summarize the Module lifecycle as follows:

Reviewer 2: Nit: 

- "Special" [``init``] is repeated 3 or 4 times here (in the `apply` context), so maybe rephrase it a little.
- "We understand" is a presumption - maybe use "you learned about"

e.g.

```suggestion
Now that you have learned about ``init`` (a special case of ``apply``), let's analyze what ``.apply(...)`` does in more detail. In fact, most of the complexity of Modules resides in the ``apply`` method. The "Module lifecycle" consists of constructing and ``apply``ing a Module. We can summarize the Module lifecycle as follows:
```

#. We construct `mlp = MLP(hidden_size=5, out_size=3)`, such that `mlp.hidden_size=5` and `mlp_out_size=3`.

> #. Then, call ``mlp.apply`` which performs the following:

Reviewer 2: Nit: this can be shortened a little 

e.g.

```suggestion
#. Then, call ``mlp.apply``, which:
```

   #. Makes a clone of ``mlp``, let's call it `mlp_copy`.

   #. Calls `mlp_copy.setup()`.

   #. Returns the output of ``mlp_copy.__call__()`` and optionally the variable collections that were specified as mutable using the keyword argument ``mutable=``.

> You might be surprised that the lifecycle includes cloning the Module instance. You might be surprised that the lifecycle includes cloning the Module instance. In short, this is done to ensure that `apply` can be treated as a pure function (i.e., if you pass the same arguments in, it will return the same outputs). This will be explained in more detail later in the  :ref:`Top-level Modules` section.

Reviewer 2: Nit: repeating sentences.

```suggestion
Notice that the lifecycle includes cloning the Module instance. This is done to ensure that `apply` can be treated as a pure function (i.e., if you pass the same arguments in, it will return the same outputs). You will learn about this in more detail later in the  :ref:`Top-level Modules` section.
```

(+ some minor refactoring)

      elif mode == "decode":
        return decoder(x)

.. testcode::
  :hide:

  def init_fn(mdl):
    x = jax.numpy.ones((3, 2))
    z = mdl(x, "encode")
    return mdl(z, "decode")

  mdl = CorrectModule()
  vars = nn.init(init_fn, mdl)(random.PRNGKey(0))
  assert vars["params"]["Dense_0"]["kernel"].shape == (2, 8)
  assert vars["params"]["Dense_1"]["kernel"].shape == (8, 4)

In the above example the construction order is fixed. After construction the submodules can be used in an arbitrary order.

.. note::
   compact modules show a strong resemblance to `React hooks <https://reactjs.org/docs/hooks-custom.html>`_.

Top-level Modules
*****************

When a Module instance is created at the "top-level", it will be in an "unbound" state - that is, it has no variables attached. "Top-level" means it is not constructed as a sub-Module inside another Module class. Apart from calling ``init`` and ``apply``, there is not much you can do with an unbound Module. Note also that ``setup`` is not called on unbound Modules, so you can really only access the construction arguments. Refer to the :ref:`Future work` section to learn how this might change in the future.

> Why are top-level modules always unbound?

Reviewer 2: Nit: maybe

```suggestion
Why are top-level Modules always unbound?
```

Cloning
===============================================

To make this approach work reliably we need well-defined cloning behavior. Rather than relying on a complex nested cloning procedure like Python's ``deepcopy``, Flax enforces that a ``Module`` is exactly defined by its construction arguments. Therefore cloning a Module reduces to calling the constructor with its original construction arguments. Because ``Module`` acts as an immutable dataclass, the construction arguments are mapped directly to instance attributes. Non-construction attributes that are computed in ``setup`` or ``__post_init__`` should also depend only on the construciton arguments to ensure a well-defined clone.

Bind
===============================================

> Sometimes it's useful to have a bound, top-level Module without having to wrap the code in a function. For example: to interact with a Module inside a Jupyter notebook. The `bind <https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=bind#flax.linen.Module.bind>`_ method returns a bound clone with an unlimited lifetime. The downside of this is that you cannot combine it with JAX transformations or integrate it into a vanilla JAX codebase that expects stateless code. For example, `Optax <https://github.com/deepmind/optax>`_ can optimze a Pytree of parameters but it cannot direclty optimize a bound ``Module`` instance created with ``.bind`` (because that's not a Pytree). Thus, you cannot combine the ``bind`` API with a functional optimizer API like Optax.

Reviewer 3: direclty -> directly

Setup
**********

The ``setup`` method is often used like the constructor hook (`__init__`) in normal Python classes. However, for more advanced use cases it's good to realize that it is not quite the same as a constructor.

> ``setup`` is only called after a Module becomes bound. Normally, this is not an issue because most Modules are bound (almost) immediately (as part of ``init`` and ``apply``). Inside ``setup``, sub-modules become bound when they are assigned to an attribute. Inside an ``nn.compact`` decorated method, sub-modules are bound immediately when constructed. As explained in the previous section, top-level Modules are never bound and thus setup is not called when they are constructed. This means you cannot access attributes assigned in setup from an unbound, top-level module.

Reviewer 4: When does the Module become bound? I don't see it getting mentioned in LifeCycle section or anywhere before this Setup section.

Author: In this paragraph it says modules become bound when the are constructed in compact or assigned to an attribute in self. In the section on top-level modules it's explained that the top-level Module itself is always unbound and init/apply operate on a clone that is bound
